### ISSUE TITLE
First pass authentication failing, second pass same authentication method working w/new value.

### ISSUE THREAD

**Issue by WebDesignerlol** [goauthentik/authentik#1740]
**Describe your question/**
I have an issue with authentication not being authorized on the first pass.
Occasionally, my TOTP authentication and security key was not authorized on first pass. Typically it fails just once then it authorizes on the second round.
It will not authorize using the same TOTP password, and my security key, even though setup properly while logged in, won't work usually if it fails this first pass fail.

**Relevant infos**
Docker on 2nd server, Nginx-Proxy with forward-auth and upstream modules enabled on first server.

**Version and Deployment (please complete the following information):**
 - authentik version: 2021.10.2

**Additional context**
I believe it maybe an error with establishing connection to Authentik server on first pass. I could be wrong, but it appears once this connection is live, it then authenticates properly on second pass.

**Comment by BeryJu** [comment]
So you're not the first to report an issue like this, there have been issues (mostly when people are using OAuth Sources), where they'll go through the entire flow, but then aren't logged in at the end. 

There have been a few attempts at debugging this, it seems to be caused by a race condition somewhere.

The main issue is of course that its not easy to reproduce

**Comment by WebDesignerlol** [comment]
I can reliably reproduce it every time now. I have a login with Windows Hello as my key, it ignores the first value, and fails if I keep selecting "Retry". If I select "Return to Device Picker", and select the same original authentication method, it will accept it second pass. Not sure if there are any log files I can give you to help with this.

**Comment by BeryJu** [comment]
I haven't tried it with Windows Hello, but I'll try to reproduce it with another WebAuthn device. If I can't I'll get a windows Hello camera and see if I can reproduce it like that.

But thanks for the further information, so it's only giving you issues at the authenticator validation stage? Because then the issue I referred to above is actually not relevant at all.

**Comment by WebDesignerlol** [comment]
Yes, just the verification stage. The issue appears for me using Windows Hello, using fingerprint scanner on mobile and TOTP as well. It seems to cache the failed response using TOTP and I have to wait until the next TOTP code renews and use that otherwise the correct code won't work as it appears to have been marked "failed". For the Windows Hello and the fingerprint on android, it won't work if I select "Retry authentication", it will only work if I select "Return to device picker". Not sure if selecting "Return to device picker" resets something in the code that might be useful to fixing this.

**Comment by BeryJu** [comment]
Interesting, I've tested with a combination of TOTP, static tokens and a Google Titan Key (instead of windows hello), and every authenticator works first try, regardless if I have all of them configured or only one.

**Comment by WebDesignerlol** [comment]
Both servers tested with SSH access and both sync up with correct current time values. Also, client is syncing with both above servers. Not sure if this is perhaps a flow issue.

**Comment by WebDesignerlol** [comment]
I have figured it out. It was a flow issue. I am not sure if this is default flow behavior or I accidentally set this up.
I had the "Default Authentication Flow" enabled as such:

- 10 | default-authentication-identification | Identification Stage | Edit Stage Edit Binding
- 20 | default-authentication-password | Password Stage | Edit Stage Edit Binding

////////////////////////// Problematic flow below
- 30 | default-authentication-flow-mfa | Authenticator Validation Stage | Edit Stage Edit Binding
//settings were "Evaluate on plan" and "Retry returns the error message and similar challenge" "Any policy should match"
////////////////////////// Problematic flow above 

- 30 | default-authentication-mfa-validation | Authenticator Validation Stage | Edit Stage Edit Binding
- 100 | default-authentication-login

I fixed this by deleting that flow and only retaining my default-authentication-mfa-validation flow. This seems to work well.

I believe this was stock. I could have messed it up initially but I don't recall doing that.

In any case this is solved with this, unless this presents a problem that I am not thinking through currently.
Please let me know if this seems an acceptable fix, and I apologize if this was caused by my own misconfiguration.

**Comment by BeryJu** [comment]
Oh, this makes sense now. Before 2021.10, to use MFA, an Admin would have to create a `Authenticator Validation Stage` by hand and bind it (which is what `default-authentication-flow-mfa` is in your case).

In 2021.10.1 I added a default stage to do that, to allow MFA out of the box without any custom flows, which is what `default-authentication-mfa-validation` is. 

I am a bit confused as the migration should check if theres already a validation stage bound, but apparently that didn't happen. Did you change the slug of your flow?

(Also I did order a lenovo 500 FHD to debug with windows hello for this, guess I won't actually need it)
